src/graphic_model/layouters/lomse_horizontal_layer.cpp

```diff
@@ -0,0 +1,234 @@
+    , m_xAbsRight(xAbsRight)
+{
+}
+
+//---------------------------------------------------------------------------------------
+void HorizontalLayer::add_shape(GmoShape* pShape)
+{
+    add_shape(pShape, pShape->get_left(), pShape->get_right());
+}
+
+//---------------------------------------------------------------------------------------
+void HorizontalLayer::add_shape(GmoShape* pShape, LUnits xLeft, LUnits xRight)
+{
+    //erase intersecting intervals
+    auto it = m_intervals.upper_bound(xLeft);
+    while (it != m_intervals.end())
+    {
+        OccupiedSpaceInterval& interval = it->second;
+
+        if (interval.xRight > xRight)
+        {
+            if (interval.xLeft < xRight)
+                m_intervals.erase(it);
+            break;
+        }
+
+        m_intervals.erase(it);
+        it = m_intervals.upper_bound(xLeft);
+    }
+
+    //add new interval
+    OccupiedSpaceInterval addedInterval { pShape, xLeft, xRight };
+    m_intervals.emplace(xRight, addedInterval);
+}
+
+//---------------------------------------------------------------------------------------
+void HorizontalLayer::align_shape_base_lines(LUnits maxAlignDistance, VerticalProfile* pVProfile, int idxStaff, bool fDown)
+{
+    if (m_intervals.empty())
+        return;
+
+    auto itGroupBegin = m_intervals.begin();
+    LUnits yBaseline = itGroupBegin->second.pShape->get_baseline_y();
+    LUnits xPrev = itGroupBegin->second.xRight;
+
+    for (auto it = itGroupBegin; it != m_intervals.end(); ++it)
+    {
+        const OccupiedSpaceInterval& interval = it->second;
+        GmoShape* pShape = interval.pShape;
+        const LUnits yShapeBaseline = pShape->get_baseline_y();
+
+        if (interval.xLeft - xPrev > maxAlignDistance)
+        {
+            set_base_line_for_range(itGroupBegin, it, yBaseline, pVProfile, idxStaff);
+            itGroupBegin = it;
+            yBaseline = yShapeBaseline;
+        }
+        else if (fDown)
+        {
+            if (yShapeBaseline > yBaseline)
+                yBaseline = yShapeBaseline;
+        }
+        else
+        {
+            if (yShapeBaseline < yBaseline)
+                yBaseline = yShapeBaseline;
+        }
+
+        xPrev = interval.xRight;
+    }
+
+    //align the last group
+    set_base_line_for_range(itGroupBegin, m_intervals.end(), yBaseline, pVProfile, idxStaff);
+}
+
+//---------------------------------------------------------------------------------------
+void HorizontalLayer::set_base_line_for_range(IntervalIterator rangeBegin, IntervalIterator rangeEnd,
+                                              LUnits yBaseline, VerticalProfile* pVProfile, int idxStaff)
+{
+    for (auto it = rangeBegin; it != rangeEnd; ++it)
+    {
+        GmoShape* pShape = it->second.pShape;
+        const LUnits yShift = yBaseline - pShape->get_baseline_y();
+
+        if (yShift)
+        {
+            pShape->shift_origin(USize(0, yShift));
+            pVProfile->update(pShape, idxStaff);
+        }
```

Comment:
A comment, for your information, just in case.

VerticalProfile was designed considering that once a shape is added to it, the shape position will not change. When a shape is added to VerticalProfile its boundaries are included in the profile. If the position of the shape is later modified, a new invocation to `VerticalProfile::update()` does not remove the old computed positions, just includes the new ones. This will imply that unoccupied space could remain marked as occupied.

But, in this method, the shift is always in vertical, up for the above HorizontalLayer and down for the bellow HorizontalLayer. Thus updating VerticalProfile does the right thing, and there is no problem. Perhaps VerticalProfile should be improved for supporting updates when a shape is moved when overlaps or by other reason (I take note).

Reply:
Yes, the shift here is always vertical and always away from a staff (at least if shapes are added to the correct layers). I don't actually think it is necessary now to support horizontal displacement in `VerticalProfile`, because horizontal placement of shapes is currently always decided before adding them to the vertical profile. If some change in the engraving algorithm requires that then this can be implemented along with that change.

Reply:
Thank you. Hope this is not necessary as in that case the VerticalProfile class will have to be totally  redesigned!